### PR TITLE
feat: storage trie implementation

### PR DESCRIPTION
**Pull request by dryajov**

### CODE REVIEW

**Comment by hermanjunge** [comment]
@dryajov while in the cold importer it was necessary to create a dummy IPLD raw object, as we are using a stripped down version of everything, here you just need to build the right CID for the `codehash` using a function like this

```go
func codeHashCid(rawdata []byte) *cid.Cid {
	c, err := cid.Prefix{
		Codec:    0x55,
		Version:  1,
		MhType:   mh.KECCAK_256,
		MhLength: -1,
	}.Sum(rawdata)
	if err != nil {
		panic(err)
	}

	return c
}
```

File: eth_account_snapshot.go [hunk 1]
```diff
@@ -89,11 +89,11 @@ func (as *EthAccountSnapshot) Resolve(p []string) (interface{}, []string, error)
 	case "balance":
 		return as.Balance, nil, nil
 	case "codeHash":
-		return fmt.Sprintf("0x%x", as.CodeHash), nil, nil
+		return &node.Link{Cid: keccak256ToCid(MEthCodeHash, as.CodeHash)}, nil, nil
 	case "nonce":
 		return as.Nonce, nil, nil
 	case "root":
-		return fmt.Sprintf("0x%x", as.Root), nil, nil
+		return &node.Link{Cid: keccak256ToCid(MEthStorageTrie, as.Root)}, nil, nil
```
Comment (dryajov) [inline eth_account_snapshot.go]: @hermanjunge not sure if it's correct. Seems like the storage root is not actually keccak encoded, only the keys are. I think this needs reverting to `rawdataToCid`?

Comment (hermanjunge) [reply eth_account_snapshot.go]: It is encoded. Each key is `keccak256(val)`

Take for example this very test file you are adding `eth-storage-trie-rlp-000dd0`, its contents, if you `xxd -p -c1000` it, are

```
f85180808080a07a6b50fdf61a20ef57cfd22a051c3bfc21185852b7f6a86dac72ac44c2b2291e80808080808080a02db9634453975816c20676cdaca000d9da7db78f947eeaaaea53db1a3e18f94380808080
```

Which encoded to bytes, and hashed correspond to

```
000dd0b4777a15d7dd300ff5c8d3eb19ec0038c8f00958494863c7f071847161
```

Comment (hermanjunge) [reply eth_account_snapshot.go]: Maybe your concern is on the merkle path. Which is correct. In the state trie, you do `keccak256(address)` to reach a leave, whereas in the storage trie, you don´t.

**Comment by hermanjunge** [comment]
@dryajov It would be _super nice_ if we can add documentation in `plugin/README.md` on this new feature 😄

**Comment by dryajov** [comment]
Absolutely! Going to do it first thing!

**Comment by dryajov** [comment]
@hermanjunge not sure what you mean here - https://github.com/ipfs/go-ipld-eth/pull/11#issuecomment-337655765? I think -https://github.com/dryajov/go-ipld-eth/blob/ff0f691684ac73c4c7f4fac895ca0317b80f5ba5/common.go#L33, already takes care of that passing the `0x55` as codec?

**Comment by hermanjunge** [comment]
Are we ready for review @dryajov ?

**Comment by dryajov** [comment]
@hermanjunge yes, this is ready.

**Comment by hermanjunge** [comment]
@dryajov The docs are a "nice to have". Are we including them in this PR, or in a further one?

**Comment by dryajov** [comment]
@hermanjunge I'll add them here :) tho if we really want this now, we can merge as is.

**Review by ghost — APPROVED** [review]
LGTM. I'm taking over the docs of storage trie.

**Pull request merged**